# [과제 26] 와인 등급 맞추기

In [143]:
from keras.models import Sequential, load_model
from keras.layers import Dense
from keras.callbacks import ModelCheckpoint, EarlyStopping
from keras.utils import np_utils
from sklearn.model_selection import train_test_split

import pandas as pd
import numpy
import os
import tensorflow as tf 

In [2]:
seed = 0
numpy.random.seed(seed)
tf.set_random_seed(seed)

In [15]:
df = pd.read_csv("../dataset/wine.csv", header=None)
df.head()

0     1     2    3      4     5     6       7     8     9    10  11  12
0   7.4  0.70  0.00  1.9  0.076  11.0  34.0  0.9978  3.51  0.56  9.4   5   1
1   7.8  0.88  0.00  2.6  0.098  25.0  67.0  0.9968  3.20  0.68  9.8   5   1
2   7.8  0.76  0.04  2.3  0.092  15.0  54.0  0.9970  3.26  0.65  9.8   5   1
3  11.2  0.28  0.56  1.9  0.075  17.0  60.0  0.9980  3.16  0.58  9.8   6   1
4   7.4  0.70  0.00  1.9  0.076  11.0  34.0  0.9978  3.51  0.56  9.4   5   1

In [303]:
dataset = df.values
X = dataset[:, 0:11] # 12번 변수를 제와한 11개 변수 선택 
Y = dataset[:, 11]
Y_encoded = np_utils.to_categorical(Y)  # 원-핫 인코딩 

In [229]:
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size = 0.3, random_state=seed)

Y_train_encoded = np_utils.to_categorical(Y_train)
Y_test_encoded = np_utils.to_categorical(Y_test)

In [243]:
X_train.shape

(4547, 11)

## **[모델 1] : 64 x 32 x 16 x 10**

In [210]:
model_1 = Sequential()
model_1.add(Dense(64,  input_dim = 11, activation='relu'))
model_1.add(Dense(32, activation='relu'))
model_1.add(Dense(16, activation='relu'))
model_1.add(Dense(10, activation='softmax'))  # 다중 분류 모델을 위한 소프트멕스 활성화 함수 

In [211]:
model_1.compile(loss='categorical_crossentropy',  # 다중 분류 모델에 사용하는 로스 함수 
              optimizer = 'adam',
              metrics=['accuracy'])

## 자동중단을 통해 최적의 모델 찾기 

In [251]:
# 모델 저장 폴더 설장 
MODEL_DIR = './model_1/'

if not os.path.exists(MODEL_DIR):
    os.mkdir(MODEL_DIR)

In [252]:
# 모델 업데이트 및 저장 
modelpath="./model_1/{epoch:02d}-{val_loss:.4f}.hdf5"
checkpointer = ModelCheckpoint(filepath=modelpath, monitor='val_loss', verbose=1, save_best_only=True)

In [253]:
# 학습 자동 중단 설정 
early_stopping_callback = EarlyStopping(monitor = 'val_loss', patience = 200)

In [254]:
model_1.fit(X, Y_encoded, validation_split=0.3, epochs=2000, batch_size=100, verbose=0,
           callbacks=[early_stopping_callback, checkpointer])


Epoch 00001: val_loss improved from inf to 1.09324, saving model to ./model_1/01-1.0932.hdf5

Epoch 00002: val_loss improved from 1.09324 to 1.06983, saving model to ./model_1/02-1.0698.hdf5

Epoch 00003: val_loss did not improve from 1.06983

Epoch 00004: val_loss did not improve from 1.06983

Epoch 00005: val_loss did not improve from 1.06983

Epoch 00006: val_loss did not improve from 1.06983

Epoch 00007: val_loss did not improve from 1.06983

Epoch 00008: val_loss did not improve from 1.06983

Epoch 00009: val_loss did not improve from 1.06983

Epoch 00010: val_loss did not improve from 1.06983

Epoch 00011: val_loss did not improve from 1.06983

Epoch 00012: val_loss did not improve from 1.06983

Epoch 00013: val_loss did not improve from 1.06983

Epoch 00014: val_loss did not improve from 1.06983

Epoch 00015: val_loss improved from 1.06983 to 1.05778, saving model to ./model_1/15-1.0578.hdf5

Epoch 00016: val_loss did not improve from 1.05778

Epoch 00017: val_loss did not imp


Epoch 00156: val_loss did not improve from 1.03863

Epoch 00157: val_loss did not improve from 1.03863

Epoch 00158: val_loss did not improve from 1.03863

Epoch 00159: val_loss did not improve from 1.03863

Epoch 00160: val_loss did not improve from 1.03863

Epoch 00161: val_loss did not improve from 1.03863

Epoch 00162: val_loss did not improve from 1.03863

Epoch 00163: val_loss did not improve from 1.03863

Epoch 00164: val_loss did not improve from 1.03863

Epoch 00165: val_loss did not improve from 1.03863

Epoch 00166: val_loss did not improve from 1.03863

Epoch 00167: val_loss did not improve from 1.03863

Epoch 00168: val_loss did not improve from 1.03863

Epoch 00169: val_loss did not improve from 1.03863

Epoch 00170: val_loss did not improve from 1.03863

Epoch 00171: val_loss did not improve from 1.03863

Epoch 00172: val_loss did not improve from 1.03863

Epoch 00173: val_loss did not improve from 1.03863

Epoch 00174: val_loss did not improve from 1.03863

Epoch 00175

In [255]:
del model_1 
model_1 = load_model('model_1/65-1.0386.hdf5')
print("\n Test Accuracy: %.4f" % (model_1.evaluate(X, Y_encoded)[1]))

6497/6497 [==============================] - 1s 103us/step

 Test Accuracy: 0.5447


- [모델 1]의 테스트 셋  정확도 0.5447

## **[모델 2] 256 x 128 x 64 x 32 x 16 x 10**

In [259]:
model_2 = Sequential()

model_2.add(Dense(256,  input_dim = 11, activation='relu'))
model_2.add(Dense(128, activation='relu'))
model_2.add(Dense(64, activation='relu'))
model_2.add(Dense(32, activation='relu'))
model_2.add(Dense(16, activation='relu'))
model_2.add(Dense(10, activation='softmax'))  # 다중 분류 모델을 위한 소프트멕스 활성화 함수 

In [266]:
model_2.compile(loss='categorical_crossentropy',  # 다중 분류 모델에 사용하는 로스 함수 
              optimizer = 'adam',
              metrics=['accuracy'])

In [267]:
# 모델 저장 폴더 설장 
MODEL_DIR = './model_2/'

if not os.path.exists(MODEL_DIR):
    os.mkdir(MODEL_DIR)
    
# 모델 업데이트 및 저장 
modelpath="./model_2/{epoch:02d}-{val_loss:.4f}.hdf5"
checkpointer = ModelCheckpoint(filepath=modelpath, monitor='val_loss', verbose=1, save_best_only=True)

# 학습 자동 중단 설정 
early_stopping_callback = EarlyStopping(monitor = 'val_loss', patience = 100)

In [274]:
model_2.fit(X, Y_encoded, validation_split=0.3, epochs=2000, batch_size=200, verbose=0,
           callbacks=[early_stopping_callback, checkpointer])


Epoch 00001: val_loss did not improve from 1.53684

Epoch 00002: val_loss did not improve from 1.53684

Epoch 00003: val_loss did not improve from 1.53684

Epoch 00004: val_loss did not improve from 1.53684

Epoch 00005: val_loss did not improve from 1.53684

Epoch 00006: val_loss did not improve from 1.53684

Epoch 00007: val_loss did not improve from 1.53684

Epoch 00008: val_loss did not improve from 1.53684

Epoch 00009: val_loss did not improve from 1.53684

Epoch 00010: val_loss did not improve from 1.53684

Epoch 00011: val_loss did not improve from 1.53684

Epoch 00012: val_loss did not improve from 1.53684

Epoch 00013: val_loss improved from 1.53684 to 1.53592, saving model to ./model_2/13-1.5359.hdf5

Epoch 00014: val_loss did not improve from 1.53592

Epoch 00015: val_loss did not improve from 1.53592

Epoch 00016: val_loss did not improve from 1.53592

Epoch 00017: val_loss did not improve from 1.53592

Epoch 00018: val_loss did not improve from 1.53592

Epoch 00019: val_

In [276]:
del model_2 
model_2 = load_model('model_2/13-1.5359.hdf5')
print("\n Test Accuracy: %.4f" % (model_2.evaluate(X, Y_encoded)[1]))

6497/6497 [==============================] - 1s 126us/step

 Test Accuracy: 0.6137


- [모델 2]의 예측 정확도 0.6137
- [모델 2]에서 은닉층을 더 많이 생성한 결과 [모델 1]의 정확도(0.5447)보다 0.069%p 증가한 것으로 나타났다.


## **[모델 3] 55 x 44 x 33 x 22 x 10**

In [288]:
model_3 = Sequential()

model_3.add(Dense(11, input_dim = 11, activation='relu'))
model_3.add(Dense(64, activation='relu'))
model_3.add(Dense(32, activation='relu'))
model_3.add(Dense(10, activation='softmax'))  # 다중 분류 모델을 위한 소프트멕스 활성화 함수 

In [289]:
model_3.compile(loss='categorical_crossentropy',  # 다중 분류 모델에 사용하는 로스 함수 
              optimizer = 'adam',
              metrics=['accuracy'])

In [292]:
# 모델 저장 폴더 설장 
MODEL_DIR = './model_3/'

if not os.path.exists(MODEL_DIR):
    os.mkdir(MODEL_DIR)
    
# 모델 업데이트 및 저장 
modelpath="./model_3/{epoch:02d}-{val_loss:.4f}.hdf5"
checkpointer = ModelCheckpoint(filepath=modelpath, monitor='val_loss', verbose=1, save_best_only=True)

# 학습 자동 중단 설정 
early_stopping_callback = EarlyStopping(monitor = 'val_loss', patience = 100)

In [293]:
model_3.fit(X, Y_encoded, validation_split=0.3, epochs=2000, batch_size=200, verbose=0,
           callbacks=[early_stopping_callback, checkpointer])


Epoch 00001: val_loss improved from inf to 1.06097, saving model to ./model_3/01-1.0610.hdf5

Epoch 00002: val_loss did not improve from 1.06097

Epoch 00003: val_loss did not improve from 1.06097

Epoch 00004: val_loss did not improve from 1.06097

Epoch 00005: val_loss did not improve from 1.06097

Epoch 00006: val_loss did not improve from 1.06097

Epoch 00007: val_loss did not improve from 1.06097

Epoch 00008: val_loss did not improve from 1.06097

Epoch 00009: val_loss did not improve from 1.06097

Epoch 00010: val_loss did not improve from 1.06097

Epoch 00011: val_loss did not improve from 1.06097

Epoch 00012: val_loss did not improve from 1.06097

Epoch 00013: val_loss did not improve from 1.06097

Epoch 00014: val_loss did not improve from 1.06097

Epoch 00015: val_loss did not improve from 1.06097

Epoch 00016: val_loss did not improve from 1.06097

Epoch 00017: val_loss did not improve from 1.06097

Epoch 00018: val_loss did not improve from 1.06097

Epoch 00019: val_loss

In [294]:
del model_3 
model_3 = load_model('model_3/01-1.0610.hdf5')
print("\n Test Accuracy: %.4f" % (model_3.evaluate(X, Y_encoded)[1]))

6497/6497 [==============================] - 1s 142us/step

 Test Accuracy: 0.5446


## **[모델 4]회귀식으로 풀어보기 64 x 32 x 16 x 1**

In [297]:
model_4 = Sequential()

model_4.add(Dense(64, input_dim = 11, activation='relu'))
model_4.add(Dense(32, activation='relu'))
model_4.add(Dense(16, activation='relu'))
model_4.add(Dense(1))

In [298]:
model_4.compile(loss='mean_squared_error', optimizer='adam')

In [314]:
# 모델 저장 폴더 설장 
MODEL_DIR = './model_4/'

if not os.path.exists(MODEL_DIR):
    os.mkdir(MODEL_DIR)
    
# 모델 업데이트 및 저장 
modelpath="./model_4/{epoch:02d}-{val_loss:.4f}.hdf5"
checkpointer = ModelCheckpoint(filepath=modelpath, monitor='val_loss', verbose=1, save_best_only=True)

# 학습 자동 중단 설정 
early_stopping_callback = EarlyStopping(monitor = 'val_loss', patience = 100)

In [315]:
model_4.fit(X_train, Y_train, validation_split=0.3, epochs=2000, batch_size=10, verbose=0,
           callbacks=[early_stopping_callback, checkpointer])


Epoch 00001: val_loss improved from inf to 0.46961, saving model to ./model_4/01-0.4696.hdf5

Epoch 00002: val_loss did not improve from 0.46961

Epoch 00003: val_loss did not improve from 0.46961

Epoch 00004: val_loss improved from 0.46961 to 0.46475, saving model to ./model_4/04-0.4647.hdf5

Epoch 00005: val_loss improved from 0.46475 to 0.46097, saving model to ./model_4/05-0.4610.hdf5

Epoch 00006: val_loss did not improve from 0.46097

Epoch 00007: val_loss did not improve from 0.46097

Epoch 00008: val_loss did not improve from 0.46097

Epoch 00009: val_loss did not improve from 0.46097

Epoch 00010: val_loss did not improve from 0.46097

Epoch 00011: val_loss did not improve from 0.46097

Epoch 00012: val_loss did not improve from 0.46097

Epoch 00013: val_loss did not improve from 0.46097

Epoch 00014: val_loss did not improve from 0.46097

Epoch 00015: val_loss did not improve from 0.46097

Epoch 00016: val_loss did not improve from 0.46097

Epoch 00017: val_loss did not imp

In [316]:
del model_4 
model_4 = load_model('model_4/05-0.4610.hdf5')

In [330]:
# 예측 값과 실제 값의 비교

Y_prediction = model_4.predict(X).flatten()

In [333]:
for i in range(10):
    real = Y_test[i]
    prediction = Y_prediction[i]
    print("실제 등급 : {0}, 예상 등급 : {1:.4}".format(real, round(prediction)))

실제 등급 : 6.0, 예상 등급 : 5.0
실제 등급 : 6.0, 예상 등급 : 5.0
실제 등급 : 6.0, 예상 등급 : 5.0
실제 등급 : 5.0, 예상 등급 : 5.0
실제 등급 : 8.0, 예상 등급 : 5.0
실제 등급 : 5.0, 예상 등급 : 5.0
실제 등급 : 6.0, 예상 등급 : 5.0
실제 등급 : 6.0, 예상 등급 : 5.0
실제 등급 : 6.0, 예상 등급 : 5.0
실제 등급 : 6.0, 예상 등급 : 6.0


In [335]:
print("\n Test Accuracy: %.4f" % (model_4.evaluate(X, Y)))

6497/6497 [==============================] - 0s 34us/step

 Test Accuracy: 0.4896


- 회귀식을 사용해 예측을 한 [모델 4]의 경우 예측값이 0.4896으로 [모델 2]에 비해 많이 감소하였다.
- 앞에서 모든 변수를 사용해서 예측을 했지만, 높은 수준의 예측률이 나타나지 않았으며 가장 높은 예측 모델은 약 61%의 예축률을 보였다.
- 따라서 더 나은 모델을 만들기 위해 와인의 등급에 양의 영향을 주는 변수를 몇 개 선택해 분석하는 방법이 필요(?)하다고 느낀다.. 